File: typescript/queue/queue.ts
interface Queue<T> {
    enqueue(item: T): void;
    dequeue(): T | undefined;
    peek(): T | undefined;
    isEmpty(): boolean;
    isFull(): boolean;
    printQueue(): void;
}

class SimpleQueue<T> implements Queue<T> {
    private readonly capacity: number;
    private items: T[] = [];

    constructor(capacity: number = 5) {
        this.capacity = capacity;
    }

    enqueue(item: T): void {
        if (this.isFull()) {
            console.log("Queue overflow");
            return;
        }
        this.items.push(item);
    }

    dequeue(): T | undefined {
        if (this.isEmpty()) {
            console.log("Queue underflow");
            return undefined;
        }
        return this.items.shift();
    }

    peek(): T | undefined {
        if (this.isEmpty()) {
            return undefined;
        }
        return this.items[0];
    }

    isEmpty(): boolean {
        return this.items.length === 0;
    }

    isFull(): boolean {
        return this.items.length >= this.capacity;
    }

    printQueue(): void {
        console.log(this.items);
    }
}

export default SimpleQueue;

File: rust/queue/src/queue.rs
use std::fmt::Debug;

pub trait Queue<T> {
    fn enqueue(&mut self, item: T);
    fn dequeue(&mut self) -> Option<T>;
    fn peek(&self) -> Option<&T>;
    fn is_empty(&self) -> bool;
    fn is_full(&self) -> bool;
    fn print_queue(&self);
}

pub struct SimpleQueue<T> {
    items: Vec<T>,
    capacity: usize,
}

impl<T> SimpleQueue<T> {
    pub fn new(capacity: usize) -> Self {
        SimpleQueue {
            items: Vec::with_capacity(capacity),
            capacity,
        }
    }
}

impl<T: Debug> Queue<T> for SimpleQueue<T> {
    fn enqueue(&mut self, item: T) {
        if self.is_full() {
            println!("Queue overflow");
            return;
        }
        self.items.push(item);
    }

    fn dequeue(&mut self) -> Option<T> {
        if self.is_empty() {
            println!("Queue underflow");
            return None;
        }
        if self.items.len() > 0 {
            return Some(self.items.remove(0));
        } else {
            return None
        }
    }

    fn peek(&self) -> Option<&T> {
        if self.is_empty() {
            return None;
        }
        self.items.first()
    }

    fn is_empty(&self) -> bool {
        self.items.is_empty()
    }

    fn is_full(&self) -> bool {
        self.items.len() >= self.capacity
    }

    fn print_queue(&self) {
        println!("Queue: {:?}", self.items);
    }
}